lib/EFI.pm

```diff
@@ -0,0 +1,570 @@
+		chr(0xFF) x 16, # GUID
+		chr(0xFF) x ($len - $ffs_hdr_len), # data
+	);
+
+	return $ffs;
+}
+
+# Generate a DEPEX section
+sub depex
+{
+	my $type = shift;
+	return unless @_;
+
+	my $section_type = $depex_type_map{$type}
+		or die "$type: DEPEX is not supported\n";
+
+	if ($_[0] eq 'TRUE')
+	{
+		# Special case for short-circuit
+		return section($section_type, chr(0x06) . chr(0x08));
+	}
+
+	my $data = '';
+	my $count = 0;
+
+	for my $guid (@_)
+	{
+		# push the guid
+		$data .= chr(0x02) . guid($guid);
+		$count++;
+	}
+
+	# AND them all together (1 minus the number of GUIDs)
+	$data .= chr(0x03) for 1..$count-1;
+	$data .= chr(0x08);
+	
+	return section($section_type, $data);
+}
+
+
+# compress a section and Wrap a GUIDed section around it
+sub compress
+{
+	# We could force 128-byte alignment for compressed sections
+	# but it doesn't seem to matter.
+	my $data = ffs_align(0, 00, @_);
+
+	my ($fh,$filename) = tempfile();
+	print $fh $data;
+	close $fh;
+
+	# -7 produces the same bit-stream as the UEFI tools
+	#my $lz = new Compress::Raw::Lzma::EasyEncoder(Preset => 7);
+	#my $lz_data;
+	#$lz->code($data, $lz_data);
+	my $lz_data = `lzma --compress --stdout -7 $filename`;
+	#printf STDERR "%d compressed to %d\n", length($data), length($lz_data);
+
+	# fixup the size field in the lzma compressed data
+	substr($lz_data, 5, 8) = write64(length $data);
+
+	# wrap the lzdata in a GUIDed section
+	my $lz_header = ''
+		. guid($lzma_guid)
+		. chr($ffs_hdr_len)  # data offset, should this be 0x14?
+		. chr(0x00)
+		. chr(0x01)  # Processing required
+		. chr(0x00)
+		;
+
+	# and replace our data with the GUID defined LZ compressed data
+	return section(GUID_DEFINED => $lz_header . $lz_data);
+}
+
+
+# Create a FV header for a given file size
+sub fv
+{
+	my $size = shift;
+	my $guid = guid("8C8CE578-8A3D-4F1C-9935-896185C32DD3");  # FFSv2
+	#my $guid = guid("5473c07a-3dcb-4dca-bd6f-1e9689e7349a"); # FFSv3 for large sections
+
+	my $fv_hdr = ''
+		. (chr(0x00) x 0x10)		# 0x00 Zero vector
+		. $guid				# 0x10
+		. write64($size)		# 0x20 length (64-bit)
+		. '_FVH'			# 0x28 signature
+		. pack("V", 0x000CFEFF)		# 0x2C attributes
+		. pack("v", $fv_hdr_len)	# 0x30 header length (32-bit)
+		. pack("v", 0x0000)		# 0x32 checksum
+		. chr(0x00)			# 0x34 reserved?
+		. chr(0x00)			# 0x35 reserved
+		. chr(0x00)			# 0x36 reserved
+		. chr(0x02)			# 0x37 version
+		. pack("V", $size / $fv_block_size) # 0x38 number blocks (32-bit)
+		. pack("V", $fv_block_size)	# 0x3C block size (32-bit)
+		. pack("V", 0)			# 0x40 number blocks (unused)
+		. pack("V", 0)			# 0x44 block size (unused)
+		;
+
+	die "FV Header length ", length $fv_hdr, " != $fv_hdr_len\n"
+		unless $fv_hdr_len == length $fv_hdr;
+
+	# update the header checksum
+	my $sum = 0;
+	for(my $i = 0 ; $i < $fv_hdr_len ; $i += 2)
+	{
+		$sum -= unpack("v", substr($fv_hdr, $i, 2));
+	}
+
+	substr($fv_hdr, 0x32, 2) = pack("v", $sum & 0xFFFF);
+
+	return $fv_hdr;
+}
+
+# Add files to an existing FV
```

Comment:
I'm planning on changing the FV creation to use the same technique as the FFS -- pass in a list of FFS to the FV constructor and it will pad them correctly (or return an error).  That will eliminate this class of error.